File: src/domain/entities/Product.ts
export class ProductId {
  constructor(public readonly value: number) {}
}

export class ProductName {
  constructor(public readonly value: string) {}
}

export class ProductPrice {
  constructor(public readonly value: number) {}
}

export class ProductCategory {
  constructor(public readonly value: string) {}
}

export class Product {
  constructor(
    public readonly id: ProductId,
    public readonly name: ProductName,
    public readonly price: ProductPrice,
    public readonly category: ProductCategory,
    public readonly description: string,
    public readonly isActive: boolean,
    public readonly imageUrl?: string,
    public readonly createdAt: Date = new Date(),
    public readonly updatedAt: Date = new Date()
  ) {}

  // Business logic methods
  isAvailable(): boolean {
    return this.isActive
  }

  hasImage(): boolean {
    return !!this.imageUrl
  }

  updatePrice(newPrice: ProductPrice): Product {
    return new Product(
      this.id,
      this.name,
      newPrice,
      this.category,
      this.description,
      this.isActive,
      this.imageUrl,
      this.createdAt,
      new Date()
    )
  }

  toggleStatus(): Product {
    return new Product(
      this.id,
      this.name,
      this.price,
      this.category,
      this.description,
      !this.isActive,
      this.imageUrl,
      this.createdAt,
      new Date()
    )
  }
}

File: SmoochoPOS-Android/app/src/main/java/com/beloop/pos/data/model/Product.kt
package com.beloop.pos.data.model

import android.os.Parcelable
import androidx.room.Entity
import androidx.room.PrimaryKey
import kotlinx.parcelize.Parcelize
import java.util.Date

@Entity(tableName = "products")
@Parcelize
data class Product(
    @PrimaryKey(autoGenerate = true)
    val id: Long = 0,
    val name: String,
    val price: Double,
    val category: String,
    val description: String? = null,
    val imageUrl: String? = null,
    val isActive: Boolean = true,
    val isAvailable: Boolean = true,
    val preparationTime: Int = 0, // in minutes
    val ingredients: List<String> = emptyList(),
    val allergens: List<String> = emptyList(),
    val nutritionalInfo: NutritionalInfo? = null,
    val tags: List<String> = emptyList(),
    val createdAt: Date = Date(),
    val updatedAt: Date = Date(),
    val lastSyncedAt: Date? = null,
    val syncStatus: SyncStatus = SyncStatus.PENDING
) : Parcelable

@Parcelize
data class NutritionalInfo(
    val calories: Int = 0,
    val protein: Double = 0.0,
    val carbs: Double = 0.0,
    val fat: Double = 0.0,
    val fiber: Double = 0.0,
    val sugar: Double = 0.0
) : Parcelable

